### PR TITLE
[JENKINS-23571] Add configuration option for cache directory location

### PR DESCRIPTION
**Pull request by mryan43**
[JENKINS-23571](https://issues.jenkins-ci.org/browse/JENKINS-23571)

This PR adds a new optional configuration option for "mercurial installations".

This new options allows users to customize the location of cached mercurial repositories on the jenkins master node when using repository caching. The default value remains unchanged ($JENKINS_HOME/hgcache)

This feature is useful for people who want to place this cache outside of $JENKINS_HOME.

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/mercurial-plugin/91)

### CODE REVIEW

File: src/main/java/hudson/plugins/mercurial/Cache.java [hunk 1]
```diff
@@ -241,12 +251,19 @@ private synchronized ReentrantLock getLockForSlaveNode(String node) {
     /**
      * Hash a URL into a string that only contains characters that are safe as directory names.
      */
-    static String hashSource(String source, StandardUsernameCredentials credentials) {
+    static String hashSource(String source, StandardUsernameCredentials credentials, String masterCacheRoot) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: `@CheckForNull`

File: src/main/java/hudson/plugins/mercurial/Cache.java [hunk 2]
```diff
@@ -241,12 +251,19 @@ private synchronized ReentrantLock getLockForSlaveNode(String node) {
...
             source += "/";
         }
         Matcher m = Pattern.compile(".+[/]([^/:]+)(:\\d+)?[/]?").matcher(source);
-        String digestible = credentials == null ? source : source + '#' + credentials.getId();
+        String digestible = source;
+        if (credentials != null){
+            digestible +=  '#' + credentials.getId();
+        }
+        if (masterCacheRoot != null){
+            digestible += "#" + masterCacheRoot.replaceAll(File.pathSeparator, "_");
```
Comment (jglick) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: You do not need to replace the `pathSeparator` here—this is just input to a digest, not part of the final directory name.

File: src/main/java/hudson/plugins/mercurial/MercurialInstallation.java [hunk 3]
```diff
@@ -64,21 +64,34 @@
...
     @Deprecated
     public MercurialInstallation(String name, String home, String executable,
             boolean debug, boolean useCaches,
             boolean useSharing, List<? extends ToolProperty<?>> properties) {
-        this(name, home, executable, debug, useCaches, useSharing, null, properties);
+        this(name, home, executable, debug, useCaches, null, useSharing, null, properties);
     }
 
-    @DataBoundConstructor public MercurialInstallation(String name, String home, String executable, boolean debug, boolean useCaches, boolean useSharing, String config, List<? extends ToolProperty<?>> properties) {
+    /** for backwards compatibility */
```
Comment (jglick) [inline src/main/java/hudson/plugins/mercurial/MercurialInstallation.java]: Could probably switch to a `@DataBoundConstructor` with just `name` and `home` and `properties` parameters, and use `@DataBoundSetter` for all the optional stuff. But cannot remember trying it in the context of a `ToolInstallation`.

File: src/test/java/hudson/plugins/mercurial/CachingCustomDirSCMTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+	private static final String CACHING_INSTALLATION = "caching-custom-dir";
+
+	@Before
+	public void setUp() throws Exception {
+		repo = tmp.getRoot();
+		j.jenkins
+				.getDescriptorByType(MercurialInstallation.DescriptorImpl.class)
+				.setInstallations(new MercurialInstallation(CACHING_INSTALLATION, "", "hg",
+						false, true, Paths.get("target", "test-data", "custom-cache-dir").toString(), false, "",
```
Comment (jglick) [inline src/test/java/hudson/plugins/mercurial/CachingCustomDirSCMTest.java]: Use `tmp.newFolder().getAbsolutePath()` instead.

File: src/main/java/hudson/plugins/mercurial/Cache.java (hunk 2 above)
Comment (mryan43) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: The intent here was to avoid environment specific data in the digest hash so that it would be the same on windows and linux for example.

File: src/test/java/hudson/plugins/mercurial/CachingCustomDirSCMTest.java (hunk 4 above)
Comment (mryan43) [inline src/test/java/hudson/plugins/mercurial/CachingCustomDirSCMTest.java]: I still need "custom-cache-dir" in the path to be able to assert that the new configuration option was really used as target path for cloning. I'll change it to new File(tmp.newFolder(),"custom-cache-dir").getAbsolutePath() if that's ok for you.

File: src/main/java/hudson/plugins/mercurial/Cache.java [hunk 5]
```diff
@@ -104,12 +108,18 @@ private synchronized ReentrantLock getLockForSlaveNode(String node) {
             if (master == null) { // Should not happen
                 throw new IOException("Cannot retrieve the Jenkins master node");
             }
-            final FilePath rootPath = master.getRootPath();
-            if (rootPath == null) {
-                throw new IOException("Cannot retrieve the root directory of the Jenkins master node");
+
+            FilePath masterCaches = null;
+            if (masterCacheRoot != null){
+                masterCaches = new FilePath(master.getChannel(), masterCacheRoot);
```
Comment (mryan43) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: How should I deal with the following 
`FilePath masterCache = masterCaches.child(hash);`
which is 9 lines lower ?

File: src/main/java/hudson/plugins/mercurial/MercurialInstallation.java (hunk 3 above)
Comment (mryan43) [inline src/main/java/hudson/plugins/mercurial/MercurialInstallation.java]: I mimicked how it was done when the "config" option was added. According to your suggestion, I suppose the constructor which doesn't take a "config" should also be removed, which would be an unrelated change to the feature I'm trying to add.

**Comment by mryan43** [comment]
I have pushed fixes regarding your comments, do you want me to squash it with the initial commit ?

**Comment by jglick** [comment]
No squashing please.

File: src/main/java/hudson/plugins/mercurial/Cache.java (hunk 5 above)
Comment (jglick) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: Sorry, I meant

```java
masterCaches = new FilePath(masterCacheRoot);
```

File: src/main/java/hudson/plugins/mercurial/Cache.java (hunk 2 above)
Comment (jglick) [inline src/main/java/hudson/plugins/mercurial/Cache.java]: OK…irrelevant since a given master is either running on Windows or it is running on Unix-ish, and that is very unlikely to change between restarts. But harmless I suppose.
